src/wasm-emscripten.cpp

```diff
@@ -196,6 +196,96 @@ void EmscriptenGlueGenerator::generateDynCallThunks() {
   }
 }
 
+struct JSCallWalker : public PostWalker<JSCallWalker> {
+  Module &wasm;
+  JSCallWalker(Module &_wasm) : wasm(_wasm) {
+    if (wasm.table.segments.size() == 0) {
+      auto emptySegment =
+          wasm.allocator.alloc<Const>()->set(Literal(uint32_t(0)));
+      wasm.table.segments.emplace_back(emptySegment);
+    }
+    std::vector<Name> tableSegmentData = wasm.table.segments[0].data;
+
+    // Check if jsCalls have already been created
+    for (unsigned i = 0; i < tableSegmentData.size(); ++i) {
+      if (tableSegmentData[i].startsWith("jsCall_")) {
+        jsCallStartIndex = i;
+        return;
+      }
+    }
+    jsCallStartIndex = wasm.table.segments[0].data.size();
+  }
+  void visitCallIndirect(CallIndirect *curr);
+
+  bool createJSCallThunks;
+  unsigned jsCallStartIndex;
+  // Function type signatures used in call_indirect instructions
+  std::set<std::string> indirectlyCallableSigs;
+};
+
+// Gather all function signatures used in call_indirect, because any of them
+// can be used to call function pointers created by emscripten's addFunction.
+void JSCallWalker::visitCallIndirect(CallIndirect* curr) {
+  indirectlyCallableSigs.insert(getSig(wasm.getFunctionType(curr->fullType)));
+}
+
+JSCallWalker getJSCallWalker(Module& wasm) {
+  JSCallWalker walker(wasm);
+  walker.walkModule(&wasm);
+  return walker;
+}
+
+void EmscriptenGlueGenerator::generateJSCallThunks(
+    unsigned numReservedFunctionPointers) {
+  if (numReservedFunctionPointers == 0)
+    return;
+
+  JSCallWalker walker = getJSCallWalker(wasm);
+  for (std::string sig : walker.indirectlyCallableSigs) {
+    // Add imports for jsCall_sig (e.g. jsCall_vi).
+    // Imported jsCall_sig functions have their first parameter as an index to
+    // the function table, so we should prepend an 'i' to parameters' signature
+    // (e.g. If the signature of the callee is 'vi', the imported jsCall_vi
+    // function would have signature 'vii'.)
+    std::string importSig = std::string(1, sig[0]) + 'i' + sig.substr(1);
```

Comment:
I didn't know that `std::string` constructor existed. Huh. This is clever (in a not-bad kind of way), it reads like python string slicing.

Reply:
Yeah what I wanted to do is `sig[0] + 'i' + sig[1:]`, but `std::string` did not have a one-argument constructor from a `char`.